beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package light
+
+import (
+	"encoding/binary"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/params"
+	"github.com/ethereum/go-ethereum/beacon/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	ErrNeedCommittee      = errors.New("sync committee required")
+	ErrInvalidUpdate      = errors.New("invalid committee update")
+	ErrInvalidPeriod      = errors.New("invalid update period")
+	ErrWrongCommitteeRoot = errors.New("wrong committee root")
+	ErrCannotReorg        = errors.New("can not reorg committee chain")
+)
+
+// CommitteeChain is a passive data structure that can validate, hold and update
+// a chain of beacon light sync committees and updates. It requires at least one
+// externally set fixed committee root at the beginning of the chain which can
+// be set either based on a CheckpointData or a trusted source (a local beacon
+// full node). This makes the structure useful for both light client and light
+// server setups.
+//
+// It always maintains the following consistency constraints:
+//   - a committee can only be present if its root hash matches an existing fixed
+//     root or if it is proven by an update at the previous period
+//   - an update can only be present if a committee is present at the same period
+//     and the update signature is valid and has enough participants.
+//     The committee at the next period (proven by the update) should also be
+//     present (note that this means they can only be added together if neither
+//     is present yet). If a fixed root is present at the next period then the
```

Comment:
I have two main concerns to challenge the design:

1. Why do we still have the trusted server assumption. I can understand servers and clients start with a trusted checkpoint, but all the later chain extension must be verified, no trust assumption is required, isn't it?

If there is some attacks and servers/clients can easily be switched to attack chain, I will say this sync committee **is wrong by design**.

2. Can you elaborate how to backward sync the committee chain? The initial external set checkpoint specifies the committee for verifying the updates belong to this period, it's unable to verify any pre-period updates.

Reply:
1. "trusted" in this context means a local beacon node that is connected to a Geth instance acting as a light server; so it is your own node which you should trust over anything else. Clients of course do not need trusted servers.
The attack vector I mentioned (bribing the supermajority of sync committee signers) is hopefully mostly theoretical but still, I feel it's best to reduce the potential chance of success (actually misleading clients) as the honest servers will still propagate the proper updates.
Btw there is another, maybe more probable scenario where it makes sense to fix roots according to the local beacon node: if the participation score of the best available update is below the default threshold setting. In this case clients can decide to slightly lower their own threshold but servers should always propagate the best available update so that clients have this option. With a fixed root at the next period, the server's `CommitteeChain` can accept the update even if its score is below the threshold.

Reply:
2. with the `historical_roots` and `state_roots` trees we can prove any old beacon state from the most recent one and the old beacon state can prove the old committee root. Btw servers will not only backward sync committees but also checkpoints from each other, so that if the client knows an older checkpoint hash it can still be served. The beacon state proofs proving execution block roots is also synced from other servers so that old exec blocks can be proven (this is what replaces the old CHTs). Since beacon state is only generated for new blocks and beacon nodes do not retain it for very long, it is a generally necessary mechanism for our use cases to retain certain parts of the old beacon states on the server side and also to be able to sync it from other servers. This is already implemented and will be a part of a future PR.